File: Seats.scala
case class Seats private(val num: Int) extends AnyVal

object Seats {
  sealed trait Error
  case class BadCount(num: Int) extends Error {
    override def toString =
      s"number of seats was $num, but must be positive"
  }

  def make(num: Int): BadCount \/ Seats = {
    if (num < 0)
      BadCount(num).left
    else
      Seats(num).right
  }
}

File: seats.rs
pub struct Seats {
  num: i32 // private
}

pub enum Error {
  BadCount(i32)
}

impl Seats {
  pub fn make(num: i32) -> Result<Seats, Error> {
    if num <= 0 {
      Err(Error::BadCount(num))
    } else {
      Ok(Seats { num: num })
    }
  }
}
